README.md
# Bookmarks and Snippets Manager for Developers & Co

Use our [**HowTo pages**](https://www.codever.land/howto) to get you going
and [start saving time and nerves](https://dev.to/ama/how-i-manage-my-dev-bookmarks-and-save-time-and-nerves-56ae) when managing
your bookmarks and code snippets. The following extensions will assist you along the way:

## Extensions

### Bookmarks

| [![Chrome](documentation/img/chrome-24.png)](https://chrome.google.com/webstore/detail/save-link-to-bookmarksdev/diofdblfhjbpgackifolmboaiccmebjb) | [![Bookmarklet](documentation/img/bookmark-24.png)](https://www.codever.land/howto/bookmarklets) |
|:---:|:---:|
| [Extension](https://chrome.google.com/webstore/detail/save-link-to-bookmarksdev/diofdblfhjbpgackifolmboaiccmebjb) | [Bookmarklet](https://www.codever.land/howto/bookmarklets) |

### Snippets

| [![Chrome](documentation/img/chrome-24.png)](https://chrome.google.com/webstore/detail/save-code-to-bookmarksdev/mibimnflfgnaealndgnamblppebdngne) | [![Firefox](documentation/img/fx-24.png)](https://addons.mozilla.org/en-GB/firefox/addon/save-code-to-bookmarks-dev/) | [![Bookmarklet](documentation/img/bookmark-24.png)](https://www.codever.land/howto/bookmarklets) | [![IntelliJ Plugin](documentation/img/intellij-24.png)](https://plugins.jetbrains.com/plugin/14456-save-to-bookmarks-dev) |
|:---:|:---:|:---:|:---:|
| [Extension](https://chrome.google.com/webstore/detail/save-code-to-bookmarksdev/mibimnflfgnaealndgnamblppebdngne) | [Extension](https://addons.mozilla.org/en-GB/firefox/addon/save-code-to-bookmarks-dev/) | [Bookmarklet](https://www.codever.land/howto/bookmarklets) | [Plugin](https://plugins.jetbrains.com/plugin/14456-save-to-bookmarks-dev) |

**Chrome extension usage example**
 [![Save to Codever chrome extension usage](documentation/gif/save-youtube-video-to-bookmarks.dev.gif)](https://www.codever.land/howto)

*******************************************************************************************************************************
> Worthy public bookmarks are published regularly on Github in a [separate repository](https://github.com/BookmarksDev/bookmarks).
*******************************************************************************************************************************

## Setup (development or self hosting)

These instructions will get you a copy of the project up and running on your local machine for development and testing purposes.

The project is developed with the MEAN stack and [Keycloak](http://www.keycloak.org/) for authentication and authorization:
![components-graph](documentation/graphviz/components-graph.png)

The project contains two parts

* [frontend](frontend) which makes up the User Interface. This uses with Angular and Angular CLI.
* [backend](backend) which is the API supporting the UI. It uses ExpressJS with MongoDB and Keycloak. See the [OpenAPI specification](https://www.codever.land/api/docs)

### Prerequisites

What you need to run this app:

* `node` and `npm` (we recommend using [NVM](https://github.com/creationix/nvm))
  * Ensure you're running Node at least (`v10.x.x`+) and NPM (`6.x.x`+)
* [nodemon](https://nodemon.io/) - `npm install -g nodemon`
* **Docker** - we recommend using [Docker Desktop](https://www.docker.com/products/docker-desktop)

> Docker and Docker-compose are currently used only for local development

### Installing (**development setup**)

#### Start MongoDB and Keycloak server

For the **first time** uncomment the following line:

```yaml
    #command: -Dkeycloak.migration.action=import -Dkeycloak.migration.provider=dir -Dkeycloak.migration.dir=/tmp/keycloak/export-import -Dkeycloak.migration.strategy=IGNORE_EXISTING
```

in the [docker-compose](docker-compose.yml) file, so that the initial Keycloak setup (realm and users) is loaded.

> For further startups of docker-compose you should comment back this line, as it starts faster

```bash
docker-compose up
```

> Use `-d` to run it in the background

#### Install and run (frontend & backend)

You want to have the backend API running first with the following commands:

```shell
# install
nvm use   #only if you are using nvm
npm install

# run
npm run frontend #to run the frontend
npm run backend #to run the backend

#alternatively run in parallel
npm start
```

or in one line `nvm use; npm install; npm start`

This starts

* the API with [nodemon](http://nodemon.io) at [http://localhost:3000/api](http://localhost:3000/api)
 and will watch for code changes in backend and automatically redeploy.
* the frontend available at [http://localhost:4200](http://localhost:4200). Login the user/password `mock/mock` configured
  for you in the initial Keycloak setup. Some bookmarks are initially loaded with your account so you can start playing with them.

> You can create your own test user, or any other extra users by following the screenshots in
> [Add a Keycloak user](documentation/keycloak/add-keycloak-user.md) to create it.

> To be able to automatically add youtube videos published date and duration to the title you need to
create a _nodemon.json_ file based on the [backend/nodemon.json.example](backend/nodemon.json.example) and your own youtube api key

## Testing

### Backend

### Integration tests

> Keycloak and mongodb have to be started as specified above (`docker-compose up` command)

Run the integration test by issuing the following command.

```bash
cd backend;
npm run integration-tests
```

A report will be generated.

## Debugging

### Frontend

#### Dev Tools in Chromium browsers

You can use the Dev Tools in Chromium based browsers (e.g. Chrome, Brave). Navigate to **Sources** and the file you want to set your break points, like in the following example:

![chrome-dev-tools-debugging](documentation/debugging/frontend/debug-frontend-chrome-dev-tools.png)

#### IntelliJ

In IntelliJ you add Javascript Debug configuration as shown below:

![Intelli-frontend-debugging](documentation/debugging/frontend/debug-frontend-intellij.png)

### Backend

#### IntelliJ / Webstorm (Node.js plugin is required)

The ``npm run debug`` starts nodemon with the `--inspect` parameter so you can attach to this process, by using the following configuration:
![nodejs-remote-debugging](documentation/debugging/backend/attach-to-nodemon-process.png)

#### Visual Studio Code

See [Node.js debugging in VS Code with Nodemon](https://github.com/microsoft/vscode-recipes/tree/master/nodemon)

## Contributing

Please read [CONTRIBUTING.md](CONTRIBUTING.md) for details on our code of conduct, and the process for submitting pull requests to us.

## Versioning

We use [SemVer](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/codeverland/codever/tags).

## Changelog

The changelog is available in [CHANGELOG.md](CHANGELOG.md)

To update the changelog execute the following commands:

```shell
nvm use
standard-changelog
```

## License

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details


CHANGELOG.md
# [3.0.0](https://github.com/codeverland/codever/compare/2.9.0...3.0.0) (2021-03-27)


### Bug Fixes

* **fe:** correct typo in menu entry in navigation.component.html ([4ffcc33](https://github.com/codeverland/codever/commit/4ffcc338a7ff67ead50efe35378aede1f3ad4909))
* **fe:** select right tab when navigation to my tags in user dashboard ([351618a](https://github.com/codeverland/codever/commit/351618a736bb35c79bc97919650c123a041a2fed))


### feature

* **codelets:** rename codelets to snippets everywhere ([dc8ef03](https://github.com/codeverland/codever/commit/dc8ef034f1ff95b71c1f9ee7b177a34088ea5c99))


### Features

* **search:** display in search bar also recent searches extra to saved searches ([421c6d7](https://github.com/codeverland/codever/commit/421c6d7a051e8a3b908a59f7e521ef2b133b67d6))
* **snippets:** Add public tagged page for snippets ([d0b23c3](https://github.com/codeverland/codever/commit/d0b23c382ac3b9649bbcf1b93b2fbf5e7e7f700d))


### BREAKING CHANGES

* **codelets:** the api paths are also changing ("snippets" instead of "codelets")



# [2.11.0](https://github.com/codeverland/codever/compare/2.9.0...2.11.0) (2021-03-06)


### Bug Fixes

* **fe:** select right tab when navigation to my tags in user dashboard ([351618a](https://github.com/codeverland/codever/commit/351618a736bb35c79bc97919650c123a041a2fed))


### Features

* **search:** display in search bar also recent searches extra to saved searches ([421c6d7](https://github.com/codeverland/codever/commit/421c6d7a051e8a3b908a59f7e521ef2b133b67d6))
* **snippets:** Add public tagged page for snippets ([d0b23c3](https://github.com/codeverland/codever/commit/d0b23c382ac3b9649bbcf1b93b2fbf5e7e7f700d))



# [2.9.0](https://github.com/codeverland/codever/compare/2.8.0...2.9.0) (2021-01-26)


### Features

* **public-snippets-page:** public snippets page and my-snippets page in dashboard ([9f6ca7b](https://github.com/codeverland/codever/commit/9f6ca7bfe3d16afdf5ed9e292db2785b6cdfbf3f))



# [2.7.0](https://github.com/codeverland/codever/compare/2.6.0...2.7.0) (2020-11-17)


### Features

* **fe:** add filter to hot-keys dialog and non-search results ([da6571d](https://github.com/codeverland/codever/commit/da6571d53d81f1afb4d1d9c71529b964af9e1ae2))



# [2.6.0](https://github.com/codeverland/codever/compare/2.5.0...2.6.0) (2020-11-15)


### Features

* **fe:** add popext option parameter when creating bookmark ([6ed31fc](https://github.com/codeverland/codever/commit/6ed31fcd9f2c22195347bfd0563a828de41c71e0))
* **fe:** always return fresh "history" ([74bae96](https://github.com/codeverland/codever/commit/74bae96964bfcf1339339dfe753d7bfec57e64a6))
* **fe:** extract add to history to new service ([521fd64](https://github.com/codeverland/codever/commit/521fd640cf59ea75274d8629650a09aa1098f59b))
* **fe:** show more entries (20) in history dialog and make it scrollable ([a9500be](https://github.com/codeverland/codever/commit/a9500be781c36089319fbce4a63cbb2d3058c47c))



# [2.5.0](https://github.com/codeverland/codever/compare/2.4.0...2.5.0) (2020-10-23)

### Change snippets to (code) snippets only visually 


# [2.4.0](https://github.com/codeverland/codever/compare/2.3.1...2.4.0) (2020-08-25)

### Performance Improvements

* **mongo:** add index on createdAt for bookmarks ([83388b4](https://github.com/codeverland/codever/commit/83388b451410f0fe6ea8f91d7dab95908470d291))
* **mongo:** add index on tags for bookmarks ([9f20cd9](https://github.com/codeverland/codever/commit/9f20cd9b925d7b03f6f638c9f2d60dfc9d0884ca))
* **update-user-data:** improve datat transfer when clicking links and bei pinned and read later ([2f2b72b](https://github.com/codeverland/codever/commit/2f2b72ba5bf1490c6eb143df0bdfb88e08267a9c))



## [2.3.2](https://github.com/codeverland/codever/compare/2.3.1...2.3.2) (2020-08-07)


### Bug Fixes

* **fe-searchbox:** close autocomplete panel when hitting ENTER ([f0a7d39](https://github.com/codeverland/codever/commit/f0a7d3944c3eb312e9ff53501c90994caf74ab58))



## [2.3.1](https://github.com/codeverland/codever/compare/2.3.0...2.3.1) (2020-08-06)


### Bug Fixes

* **fe:** trigger search when click on lupe in search bar ([0fc7cd4](https://github.com/codeverland/codever/commit/0fc7cd45c9d1496042d2ced7866b00e4622192f5))



# [2.3.0](https://github.com/codeverland/codever/compare/2.2.0...2.3.0) (2020-08-06)


### Features

* **fe:** show search domain as text for bigger screens (>1400 currently) ([709ae64](https://github.com/codeverland/codever/commit/709ae64942269dad74a402030bde7a97e0fac16a))



# [2.2.0](https://github.com/codeverland/codever/compare/2.1.0...2.2.0) (2020-08-04)


### Features

* **recent-searches:** add recent searches to user dashboard ([4c257e3](https://github.com/codeverland/codever/commit/4c257e3ac2163c32445a07dd82ba7c48a41249f0))



# [2.1.0](https://github.com/codeverland/codever/compare/2.0.4...2.1.0) (2020-08-02)


### Bug Fixes

* remove unused pipe in snippet-details.component.html ([8f94094](https://github.com/codeverland/codever/commit/8f9409432829d2cc4a53544766714ff359fc3304))


### Features

* **fe:** add "ctrl+s" hot key to focus on the searchbox ([7129921](https://github.com/codeverland/codever/commit/712992148ef378514f2255bffe7d6dc50d107c84))
* **fe:** introduce side navigation ([39dcfbc](https://github.com/codeverland/codever/commit/39dcfbcd6a6cfbf01b4c358ce824b8bf12fadba4))



## [2.0.4](https://github.com/codeverland/codever/compare/1.0.0...2.0.4) (2020-07-18)


### Bug Fixes

* **fe:** more distances between saved searches in my dashboard ([684c7be](https://github.com/codeverland/codever/commit/684c7be9963a3aeefd5de7a05ce143436d959553))
* **fe:** overflow bootrap dropdown menu in angular expansion panel ([c996aec](https://github.com/codeverland/codever/commit/c996aec0d3377ebabfc17ded70bbbeecc781e350))
* **fe-snippets:** avoid double adding snippets tags reloaded ([52285dc](https://github.com/codeverland/codever/commit/52285dcdae3305729890b8cacf2b4fe70fc5efce))

backend/src/routes/index.spec.js
var app = require('../app');
var chai = require('chai');
var request = require('supertest');

var expect = chai.expect;

describe('API Tests root', function() {
  it('should return link to swagger documentation', function(done) {
    request(app)
      .get('/api')
      .end(function(err, res) {
        expect(res.statusCode).to.equal(200);
        expect(res.text).to.equal('API Backend supporting Codever.land - <a href="/api/docs">Swagger Docs</a>');
        done();
      });
  });
});


frontend/src/app/app.component.ts
import { Component, HostListener, OnInit } from '@angular/core';

import 'styles.scss';
import { UserDataHistoryStore } from './core/user/userdata.history.store';
import { MatDialog, MatDialogConfig } from '@angular/material/dialog';
import { HotKeysDialogComponent } from './shared/dialog/history-dialog/hot-keys-dialog.component';
import { UserDataPinnedStore } from './core/user/userdata.pinned.store';
import { UserInfoStore } from './core/user/user-info.store';
import { KeycloakService } from 'keycloak-angular';
import { LoginRequiredDialogComponent } from './shared/dialog/login-required-dialog/login-required-dialog.component';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss'],
})
export class AppComponent implements OnInit {

  url = 'https://www.codever.land';
  innerWidth: any;

  userIsLoggedIn = false;
  userId: string;

  constructor(private keycloakService: KeycloakService,
              private userInfoStore: UserInfoStore,
              private userDataHistoryStore: UserDataHistoryStore,
              private userDataPinnedStore: UserDataPinnedStore,
              private historyDialog: MatDialog,
              private loginDialog: MatDialog) {
    this.innerWidth = 100;
  }

  ngOnInit(): void {
    this.keycloakService.isLoggedIn().then(isLoggedIn => {
      if (isLoggedIn) {
        this.userIsLoggedIn = true;
        this.userInfoStore.getUserInfo$().subscribe(userInfo => {
          this.userId = userInfo.sub;
        });
      }
    });
  }

  @HostListener('window:keydown.control.p', ['$event'])
  showPinned(event: KeyboardEvent) {
    if (!this.userIsLoggedIn) {
      const dialogConfig = new MatDialogConfig();

      dialogConfig.disableClose = true;
      dialogConfig.autoFocus = true;
      dialogConfig.data = {
        message: 'You need to be logged in to see the Pinned Bookmarks popup'
      };

      this.loginDialog.open(LoginRequiredDialogComponent, dialogConfig);
    } else {
      event.preventDefault();
      const dialogConfig = new MatDialogConfig();

      dialogConfig.disableClose = false;
      dialogConfig.autoFocus = true;
      dialogConfig.width = this.getRelativeWidth();
      dialogConfig.height = this.getRelativeHeight();
      dialogConfig.data = {
        bookmarks$: this.userDataPinnedStore.getPinnedBookmarks$(this.userId, 1),
        title: '<i class="fas fa-thumbtack"></i> Pinned'
      };

      const dialogRef = this.historyDialog.open(HotKeysDialogComponent, dialogConfig);
      dialogRef.afterClosed().subscribe(
        data => {
          console.log('Dialog output:', data);
        }
      );
    }
  }

  private getRelativeWidth() {
    let relativeWidth = (window.innerWidth * 80) / 100;
    if (window.innerWidth > 1500) {
      relativeWidth = (1500 * 80) / 100;
    }

    return relativeWidth + 'px';
  }

  private getRelativeHeight() {
    let relativeHeight = (window.innerHeight * 80) / 100;
    if (window.innerHeight > 1200) {
      relativeHeight = (1200 * 80) / 100;
    }

    return relativeHeight + 'px';
  }

  @HostListener('window:keydown.control.h', ['$event'])
  showHistory(event: KeyboardEvent) {
    if (!this.userIsLoggedIn) {
      const dialogConfig = new MatDialogConfig();

      dialogConfig.disableClose = true;
      dialogConfig.autoFocus = true;
      dialogConfig.data = {
        message: 'You need to be logged in to see the History Bookmarks popup'
      };

      this.loginDialog.open(LoginRequiredDialogComponent, dialogConfig);
    } else {
      event.preventDefault();
      const dialogConfig = new MatDialogConfig();

      dialogConfig.disableClose = false;
      dialogConfig.autoFocus = true;
      dialogConfig.width = this.getRelativeWidth();
      dialogConfig.height = this.getRelativeHeight();
      dialogConfig.data = {
        bookmarks$: this.userDataHistoryStore.getAllHistory$(this.userId),
        title: '<i class="fas fa-history"></i> History'
      };

      const dialogRef = this.historyDialog.open(HotKeysDialogComponent, dialogConfig);
      dialogRef.afterClosed().subscribe(
        data => {
          console.log('Dialog output:', data);
        }
      );
    }
  }
}


frontend/src/app/public/bookmarks/homepage.component.ts
import { Component, OnDestroy, OnInit } from '@angular/core';
import { Observable, Subscription } from 'rxjs';
import { Bookmark } from '../../core/model/bookmark';
import { ActivatedRoute, Router } from '@angular/router';
import { PublicBookmarksStore } from './store/public-bookmarks-store.service';
import { allTags } from '../../core/model/all-tags.const.en';
import { KeycloakService } from 'keycloak-angular';
import { UserData } from '../../core/model/user-data';
import { UserDataStore } from '../../core/user/userdata.store';
import { MatDialog } from '@angular/material/dialog';
import { environment } from '../../../environments/environment';
import { UserInfoStore } from '../../core/user/user-info.store';
import { AppService } from '../../app.service';
import { PaginationNotificationService } from '../../core/pagination-notification.service';
import { UserDataHistoryStore } from '../../core/user/userdata.history.store';
import { UserDataPinnedStore } from '../../core/user/userdata.pinned.store';
import { UserDataReadLaterStore } from '../../core/user/userdata.readlater.store';
import { UserDataWatchedTagsStore } from '../../core/user/userdata.watched-tags.store';
import { TagFollowingBaseComponent } from '../../shared/tag-following-base-component/tag-following-base.component';
import { FeedStore } from '../../core/user/feed-store.service';
import { SearchDomain } from '../../core/model/search-domain.enum';
import { MatTabChangeEvent } from '@angular/material/tabs';
import { CookieService } from '../../core/cookies/cookie.service';
import { Feedback } from '../../core/model/feedback';
import { FeedbackService } from '../feedback/feedback.service';
import iziToast, { IziToastSettings } from 'izitoast';


@Component({
  selector: 'app-public-bookmarks',
  templateUrl: './homepage.component.html',
  styleUrls: ['./homepage.component.scss']
})
export class HomepageComponent extends TagFollowingBaseComponent implements OnInit, OnDestroy {

  readonly FIRST_PAGE = 1;

  feedBookmarks$: Observable<Bookmark[]>;
  pageNavigationSubscription: Subscription;
  tags: string[] = allTags;
  userData$: Observable<UserData>;
  private userData: UserData;

  history$: Observable<Bookmark[]>;
  pinned$: Observable<Bookmark[]>;
  readLater$: Observable<Bookmark[]>;

  userIsLoggedIn = false;
  userId: string;
  userIsLoggedIn$: Promise<boolean>;

  selectedTabIndex: number;

  currentPageFeed = this.FIRST_PAGE;
  currentPageHistory = this.FIRST_PAGE;
  currentPagePinned = this.FIRST_PAGE;
  currentPageReadLater = this.FIRST_PAGE;
  callerPaginationFeed = 'feed';
  callerPaginationHistory = 'history';
  callerPaginationPinned = 'pinned';
  callerPaginationReadLater = 'read-later';

  showAcknowledgeMigrationHeader = false;

  constructor(private appService: AppService,
              private publicBookmarksStore: PublicBookmarksStore,
              private router: Router,
              private route: ActivatedRoute,
              private keycloakService: KeycloakService,
              private userDataStore: UserDataStore,
              private userDataHistoryStore: UserDataHistoryStore,
              private userDataPinnedStore: UserDataPinnedStore,
              private userDataReadLaterStore: UserDataReadLaterStore,
              private feedStore: FeedStore,
              public userDataWatchedTagsStore: UserDataWatchedTagsStore,
              public loginDialog: MatDialog,
              private userInfoStore: UserInfoStore,
              private paginationNotificationService: PaginationNotificationService,
              private cookieService: CookieService,
              private feedbackService: FeedbackService
  ) {
    super(loginDialog, userDataWatchedTagsStore);
  }

  ngOnInit(): void {
    const acknowledgedCodeverMigration = this.cookieService.readCookie('acknowledge-codever-migration');
    if (acknowledgedCodeverMigration !== 'true') {
      this.showAcknowledgeMigrationHeader = true;
    }

    const tabQueryParam = this.route.snapshot.queryParamMap.get('tab');
    const page = this.route.snapshot.queryParamMap.get('page');
    this.userIsLoggedIn$ = this.keycloakService.isLoggedIn();

    this.userIsLoggedIn$.then(isLoggedIn => {
      if (isLoggedIn) {
        this.userIsLoggedIn = true;
        this.userData$ = this.userDataStore.getUserData$();
        this.userData$.subscribe(userData => {
          this.userId = userData.userId;
          this.userData = userData;
          this.setSelectedTabIndexFromQueryParam(tabQueryParam);
          this.setCurrentPageFromQueryParam(page, this.selectedTabIndex);
          if (this.selectedTabIndex === TabIndex.Feed) {
            this.setFeedBookmarks$(true, this.currentPageFeed);
          }
        });
      } else {
        this.setSelectedTabIndexFromQueryParam(tabQueryParam);
        this.setCurrentPageFromQueryParam(page, this.selectedTabIndex);
        if (this.selectedTabIndex === TabIndex.Feed) {
          this.setFeedBookmarks$(false, this.currentPageFeed);
        }
      }

      this.listenToClickOnLogoEvent(isLoggedIn);

      this.listenToPaginationNavigationEvents(isLoggedIn);
    });

  }

  private setSelectedTabIndexFromQueryParam(tabQueryParam) {
    switch (tabQueryParam) {
      case 'history': {
        this.selectedTabIndex = TabIndex.History;
        break;
      }
      case 'pinned': {
        this.selectedTabIndex = TabIndex.Pinned;
        break;
      }
      case 'read-later': {
        this.selectedTabIndex = TabIndex.ReadLater;
        break;
      }
      default: {
        this.selectedTabIndex = TabIndex.Feed;
      }
    }
  }

  private setFeedBookmarks$(isLoggedIn: boolean, page: number) {
    if (isLoggedIn && !this.userData?.showAllPublicInFeed) {
      this.feedBookmarks$ = this.feedStore.getFeedBookmarks$(this.userId, page);
    } else {
      this.feedBookmarks$ = this.publicBookmarksStore.getRecentPublicBookmarks$(page);
    }
  }

  private setCurrentPageFromQueryParam(page: string, selectedTabIndex: number) {
    if (page) {
      switch (selectedTabIndex) {
        case TabIndex.Feed:
          this.currentPageFeed = parseInt(page, 0);
          break;
        case TabIndex.History:
          this.currentPageHistory = parseInt(page, 0);
          break;
        case TabIndex.Pinned:
          this.currentPagePinned = parseInt(page, 0);
          break;
        case TabIndex.ReadLater:
          this.currentPageReadLater = parseInt(page, 0);
      }
    }
  }

  private listenToClickOnLogoEvent(isLoggedIn: boolean) {
    this.appService.logoClicked.subscribe(logoClicked => {
      if (logoClicked) {
        this.currentPageFeed = this.FIRST_PAGE;
        this.setFeedBookmarks$(isLoggedIn, this.currentPageFeed);
      }
    });
  }

  private listenToPaginationNavigationEvents(isLoggedIn: boolean) {
    this.pageNavigationSubscription = this.paginationNotificationService.pageNavigationClicked$.subscribe(paginationAction => {
      if (paginationAction.caller === this.callerPaginationFeed && this.selectedTabIndex === TabIndex.Feed) {
        this.currentPageFeed = paginationAction.page;
        this.setFeedBookmarks$(isLoggedIn, this.currentPageFeed);
      }
      if (paginationAction.caller === this.callerPaginationHistory && this.selectedTabIndex === TabIndex.History) {
        this.currentPageHistory = paginationAction.page;
        this.history$ = this.userDataHistoryStore.getHistory$(this.userId, paginationAction.page);
      }
      if (paginationAction.caller === this.callerPaginationPinned && this.selectedTabIndex === TabIndex.Pinned) {
        this.currentPagePinned = paginationAction.page;
        this.pinned$ = this.userDataPinnedStore.getPinnedBookmarks$(this.userId, paginationAction.page);
      }
      if (paginationAction.caller === this.callerPaginationReadLater && this.selectedTabIndex === TabIndex.ReadLater) {
        this.currentPageReadLater = paginationAction.page;
        this.readLater$ = this.userDataReadLaterStore.getReadLater$(this.userId, paginationAction.page);
      }
    });
  }

  tabSelectionChanged(event: MatTabChangeEvent) {
    this.selectedTabIndex = event.index;
    if (this.userIsLoggedIn) {
      switch (event.index) {
        case TabIndex.Feed:
          this.setFeedBookmarks$(this.userIsLoggedIn, this.currentPageFeed);
          break;
        case TabIndex.History:
          this.history$ = this.userDataHistoryStore.getHistory$(this.userId, this.currentPageHistory);
          break;
        case TabIndex.Pinned:
          this.pinned$ = this.userDataPinnedStore.getPinnedBookmarks$(this.userId, this.currentPagePinned);
          break;
        case TabIndex.ReadLater:
          this.readLater$ = this.userDataReadLaterStore.getReadLater$(this.userId, this.currentPageReadLater);
          break;
      }
    }

    const queryParamsFromIndex = this.getQueryParamsForSelectedTab(this.selectedTabIndex);
    this.router.navigate([],
      {
        relativeTo: this.route,
        queryParams: {
          tab: queryParamsFromIndex.tab,
          page: queryParamsFromIndex.page
        },
        queryParamsHandling: 'merge'
      }
    );
  }

  login(selectedTab: string) {
    const options: Keycloak.KeycloakLoginOptions = {};
    options.redirectUri = `${environment.APP_HOME_URL}?tab=${selectedTab}`;
    this.keycloakService.login(options);
  }

  private getQueryParamsForSelectedTab(tabIndex: number): TabSwitchQueryParams {
    switch (tabIndex) {
      case TabIndex.History: {
        return {tab: 'history', page: this.currentPageHistory}
        break;
      }
      case TabIndex.Pinned: {
        return {tab: 'pinned', page: this.currentPagePinned};
        break;
      }
      case TabIndex.ReadLater : {
        return {tab: 'read-later', page: this.currentPageReadLater};
        break;
      }
      default: {
        return {tab: 'feed', page: this.currentPageFeed};
      }
    }
  }

  ngOnDestroy(): void {
    this.pageNavigationSubscription.unsubscribe();
  }

  searchPublicBookmarksByTag(tag: string) {
    this.router.navigate(['./search'],
      {
        queryParams: {q: '[' + tag + ']', sd: SearchDomain.PUBLIC_BOOKMARKS, page: this.FIRST_PAGE}
      });
  }

  public acknowledgeCodeverRebranding(response: string) {
    this.cookieService.createCookie('acknowledge-codever-migration', 'true', 365);
    this.showAcknowledgeMigrationHeader = false;

    const iziToastSettings: IziToastSettings = {
      title: 'Thank you for your feedback',
      timeout: 3000,
      position: 'topRight'
    }
    iziToast.success(iziToastSettings);

    const feedback: Feedback = {
      question: 'Bookmarks.dev rebranding to Codever',
      userResponse: response,
      userId: this.userId ? this.userId : null,
      userAgent: navigator.userAgent
    }

    this.feedbackService.createBookmark(feedback).subscribe();
  }
}

export interface TabSwitchQueryParams {
  tab: string;
  page: number;
}

enum TabIndex {
  Feed = 0,
  History,
  Pinned,
  ReadLater
}


docker-compose-setup/init-mongo.js
db.createUser(
  {
    user: "bookmarks",
    pwd: "secret",
    roles: [
      {
        role: "readWrite",
        db: "dev-bookmarks"
      }
    ]
  }
);


db.auth("bookmarks", "secret");

//insert initial public dev bookmarks
db.bookmarks.insert(
  [
    {
      "tags": [
        "programming",
        "blog",
        "open-source"
      ],
      "name": "Share coding knowledge – CodepediaOrg",
      "location": "https://www.codepedia.org/",
      "userId": "a7908cb5-3b37-4cc1-a751-42f674d870e1",
      "userDisplayName": "Mock",
      "language": "en",
      "description": "Coding knowledge hub, providing free educational content for professionals involved in software development. The website covers different topics and technologies with posts whose difficulty levels range from beginner to “hard-core” programming.",
      "descriptionHtml": "<p>Coding knowledge hub, providing free educational content for professionals involved in software development. The website covers different topics and technologies with posts whose difficulty levels range from beginner to “hard-core” programming.</p>",
      "publishedOn": null,
      "sourceCodeURL": "https://github.com/CodepediaOrg/codepediaorg.github.io",
      "public": true,
      "lastAccessedAt": ISODate("2019-10-23T12:22:10.480Z"),
      "likeCount": 0,
      "createdAt": ISODate("2019-10-23T12:22:10.526Z"),
      "updatedAt": ISODate("2019-10-23T12:22:10.526Z"),
      "__v": 0
    },
    {
      "tags": [
        "programming",
        "blog",
        "resources",
        "open-source"
      ],
      "name": "Bookmarks and Snippets Manager for Developers & Co",
      "location": "https://www.codever.land/",
      "userId": "a7908cb5-3b37-4cc1-a751-42f674d870e1",
      "userDisplayName": "Mock",
      "language": "en",
      "description": "Bookmarks Manager for Developers & Co",
      "descriptionHtml": "<p>Bookmarks Manager for Developers &amp; Co</p>",
      "publishedOn": null,
      "sourceCodeURL": "https://github.com/codeverland/codever",
      "public": true,
      "lastAccessedAt": ISODate("2019-10-23T12:23:53.471Z"),
      "likeCount": 0,
      "createdAt": ISODate("2019-10-23T12:23:53.486Z"),
      "updatedAt": ISODate("2019-10-23T12:23:53.486Z"),
      "__v": 0
    },
    {
      "tags": [
        "programming",
        "resource",
        "blog",
        "open-source"
      ],
      "name": "Collection of public dev bookmarks, shared with love from www.codever.land",
      "location": "https://github.com/BookmarksDev/bookmarks#readme",
      "userId": "a7908cb5-3b37-4cc1-a751-42f674d870e1",
      "userDisplayName": "Mock",
      "language": "en",
      "description": ":bookmark: :star: Collection of public dev bookmarks, shared with :heart: from www.codever.land  - CodepediaOrg/bookmarks",
      "descriptionHtml": "<p>:bookmark: :star: Collection of public dev bookmarks, shared with :heart: from www.codever.land  - CodepediaOrg/bookmarks</p>",
      "publishedOn": null,
      "sourceCodeURL": "https://github.com/BookmarksDev/bookmarks",
      "public": true,
      "lastAccessedAt": ISODate("2019-10-23T12:24:50.804Z"),
      "likeCount": 0,
      "createdAt": ISODate("2019-10-23T12:24:50.823Z"),
      "updatedAt": ISODate("2019-10-23T12:24:50.823Z"),
      "__v": 0
    },
    {
      "tags": [
        "programming",
        "resource",
        "blog",
        "open-source"
      ],
      "name": "Getting started with www.codever.land",
      "location": "https://www.codever.land/howto",
      "userId": "a7908cb5-3b37-4cc1-a751-42f674d870e1",
      "userDisplayName": "Mock",
      "language": "en",
      "description": "How to save and search bookmarks, use bookmarklets and chrome extension",
      "descriptionHtml": "<p>How to save and search bookmarks, use bookmarklets and chrome extension. Use snippets and more</p>",
      "publishedOn": null,
      "sourceCodeURL": "https://github.com/BookmarksDev/bookmarks",
      "public": true,
      "lastAccessedAt": ISODate("2020-03-23T12:24:50.804Z"),
      "likeCount": 0,
      "createdAt": ISODate("2020-03-23T12:24:50.823Z"),
      "updatedAt": ISODate("2020-03-23T12:24:50.823Z"),
      "__v": 0
    }
  ]
);

//bookmarks indexes
db.bookmarks.createIndex(
  {
    name: "text",
    location: "text",
    description: "text",
    tags: "text",
    sourceCodeURL: "text",
  },
  {
    weights: {
      name: 2,
      location: 5,
      description: 1,
      tags: 3,
      sourceCodeURL: 1
    },
    name: "full_text_search",
    default_language: "none",
    language_override: "none"
  }
);

db.bookmarks.createIndex(
  {location: 1, userId: 1},
  {
    unique: true,
    name: "unique_user_and_location"
  }
);

// snippets indexes

db.snippets.createIndex(
  {userId: 1}
);

db.snippets.createIndex(
  {
    title: "text",
    tags: "text",
    "codeSnippets.comment": "text",
    "codeSnippets.code": "text",
    sourceUrl: "text"
  },
  {
    weights: {
      title: 8,
      tags: 13,
      "codeSnippets.comment": 3,
      "codeSnippets.code": 3,
      sourceUrl: 1
    },
    name: "full_text_search",
    default_language: "none",
    language_override: "none"
  }
);
